File: 백준/Platinum/15647. 로스팅하는 엠마도 바리스타입니다/로스팅하는 엠마도 바리스타입니다.py
# 로스팅하는 엠마도 바리스타입니다
import sys
input = sys.stdin.readline

N = int(input())
total = [0] * (N+1)
childs = [0] * (N+1)
graph = [[] for _ in range(N+1)]

for _ in range(N-1):
    n1, n2, dist = map(int,input().split())
    graph[n1].append((n2, dist))
    graph[n2].append((n1, dist))

def calRoot_iterative(start):
    stack = [(start, -1, 0)]  
    post = []

    while stack:
        cur, parent, dist = stack.pop()
        post.append((cur, parent, dist))
        for neighbor, d in graph[cur]:
            if neighbor == parent:
                continue
            stack.append((neighbor, cur, d))

    for cur, parent, dist in reversed(post):
        cnt = 0
        for neighbor, d in graph[cur]:
            if neighbor == parent:
                continue
            cnt += childs[neighbor] + 1
            total[start] += total[neighbor] + (childs[neighbor] + 1) * d
        childs[cur] = cnt

def calDist_iterative(start):
    stack = [(start, -1, 0)]
    while stack:
        cur, parent, dist = stack.pop()

        if parent != -1:
            shorter = childs[cur] * dist
            longer = (N - childs[cur] - 2) * dist
            total[cur] = total[parent] + longer - shorter

        for neighbor, d in graph[cur]:
            if neighbor == parent:
                continue
            stack.append((neighbor, cur, d))

calRoot_iterative(1)
calDist_iterative(1)

for i in range(1, N+1):
    print(total[i])

File: 백준/Platinum/15647. 로스팅하는 엠마도 바리스타입니다/로스팅하는 엠마도 바리스타입니다.java
import java.io.*;
import java.util.*;

public class Main {

    static int N;
    static ArrayList<ArrayList<NodeDist>> graph;
    static int[] childs;
    static long[] total;

    static class NodeDist {
        int node;
        int dist;

        NodeDist(int node, int dist) {
            this.node = node;
            this.dist = dist;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        StringBuilder sb = new StringBuilder();

        N = Integer.parseInt(br.readLine());
        childs = new int[N+1];
        total = new long[N+1];
        graph = new ArrayList<>();

        for (int i = 0; i <= N; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < N-1; i++) {
            st = new StringTokenizer(br.readLine());

            int n1 = Integer.parseInt(st.nextToken());
            int n2 = Integer.parseInt(st.nextToken());
            int dist = Integer.parseInt(st.nextToken());

            graph.get(n1).add(new NodeDist(n2, dist));
            graph.get(n2).add(new NodeDist(n1, dist));
        }

        calRoot(1, -1, 1, 0);
        calDist(0, 1, 0);

        for (int i = 1; i < N+1; i++) {
            sb.append(total[i]).append("\n");
        }

        System.out.println(sb);

    }

    static int calRoot(int root, int prev, int cur, int dist) {

        total[root] += dist;
        int ret = 0;

        for (NodeDist n : graph.get(cur)) {
            if (n.node == prev) continue;

            ret += 1;
            ret += calRoot(root, cur, n.node, n.dist + dist);
        }

        childs[cur] = ret;

        return ret;
    }

    static void calDist(int prev, int cur, long dist) {

        long shorter = childs[cur] * dist;
        long longer = (N - childs[cur] - 2) * dist;

        total[cur] += total[prev] + (longer - shorter);

        for (NodeDist n : graph.get(cur)) {
            if (n.node == prev) continue;

            calDist(cur, n.node, n.dist);
        }
    }
}
